react: unable to use getSnapshotBeforeUpdate

I seem to be unable to use components that define `getSnapshotBeforeUpdate`, one of the new React lifecycle methods.  The DefinitelyTyped tests pass because they only _define_ the method, but the error comes when you attempt to actually use the defined component.  The error looks like:

[repo version](https://github.com/jacobwgillespie/typescript-react-reproduction)

```typescript
import * as React from 'react'

interface Props {
  number: number
}

interface Snapshot {
  number: number
}

class Demo extends React.Component<Props, {}, Snapshot> {
  getSnapshotBeforeUpdate() {
    return {number: 321}
  }

  componentDidUpdate(_prevProps: Props, _prevState: {}, snapshot?: Snapshot) {
    console.log(snapshot)
  }

  render() {
    return <strong>{this.props.number}</strong>
  }
}

export class App extends React.Component {
  render() {
    // ERROR HERE:
    return <Demo number={123} />
  }
}
```

```
src/index.tsx:30:12 - error TS2605: JSX element type 'Demo' is not a constructor function for JSX elements.
  Types of property 'getSnapshotBeforeUpdate' are incompatible.
    Type '() => { number: number; }' is not assignable to type '((prevProps: Readonly<any>, prevState: Readonly<{}>) => null) | undefined'.
      Type '() => { number: number; }' is not assignable to type '(prevProps: Readonly<any>, prevState: Readonly<{}>) => null'.
        Type '{ number: number; }' is not assignable to type 'null'.
```

Here is a repo that reproduces this issue: https://github.com/jacobwgillespie/typescript-react-reproduction.  This is using React v16.3.1, TypeScript v2.8.1, and `@types/react` v16.3.5.

It appears that TypeScript thinks that the `getSnapshotBeforeUpdate` method must always return null.  I do not understand why this is happening from looking at the code, and especially given that the `getSnapshotBeforeUpdate` method is typed via the type defs (returning an invalid snapshot does cause an error).

/cc @Kovensky @johnnyreilly 

----

- [x] I tried using the `@types/react` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.

--- a/types/react/index.d.ts
+++ b/types/react/index.d.ts
@@ -255,7 +255,7 @@ declare namespace React {
 
     // Base component for plain JS classes
     // tslint:disable-next-line:no-empty-interface
-    interface Component<P = {}, S = {}, SS = never> extends ComponentLifecycle<P, S, SS> { }
+    interface Component<P = {}, S = {}, SS = any> extends ComponentLifecycle<P, S, SS> { }
     class Component<P, S> {
         constructor(props: P, context?: any);
 
@@ -283,7 +283,7 @@ declare namespace React {
         };
     }
 
-    class PureComponent<P = {}, S = {}> extends Component<P, S> { }
+    class PureComponent<P = {}, S = {}, SS = any> extends Component<P, S, SS> { }
 
     interface ClassicComponent<P = {}, S = {}> extends Component<P, S> {
         replaceState(nextState: S, callback?: () => void): void;

--- a/types/react/test/index.ts
+++ b/types/react/test/index.ts
@@ -25,6 +25,10 @@ interface State {
     seconds?: number;
 }
 
+interface Snapshot {
+    baz: string;
+}
+
 interface Context {
     someValue?: string;
 }
@@ -51,7 +55,7 @@ declare const container: Element;
 // Top-Level API
 // --------------------------------------------------------------------------
 
-class ModernComponent extends React.Component<Props, State>
+class ModernComponent extends React.Component<Props, State, Snapshot>
     implements MyComponent, React.ChildContextProvider<ChildContext> {
     static propTypes: React.ValidationMap<Props> = {
         foo: PropTypes.number
@@ -102,6 +106,14 @@ class ModernComponent extends React.Component<Props, State>
 
     shouldComponentUpdate(nextProps: Props, nextState: State, nextContext: any): boolean {
         return shallowCompare(this, nextProps, nextState);
+    }
+
+    getSnapshotBeforeUpdate(prevProps: Readonly<Props>) {
+        return { baz: `${prevProps.foo}baz` };
+    }
+
+    componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>, snapshot: Snapshot) {
+        return;
     }
 }
 

--- a/types/react/test/tsx.tsx
+++ b/types/react/test/tsx.tsx
@@ -152,6 +152,24 @@ class ComponentWithNewLifecycles extends React.Component<NewProps, NewState, { b
 }
 <ComponentWithNewLifecycles foo="bar" />;
 
+class PureComponentWithNewLifecycles extends React.PureComponent<NewProps, NewState, { baz: string }> {
+    static getDerivedStateFromProps: React.GetDerivedStateFromProps<NewProps, NewState> = (nextProps) => {
+        return { bar: `${nextProps.foo}bar` };
+    }
+
+    getSnapshotBeforeUpdate(prevProps: Readonly<NewProps>) {
+        return { baz: `${prevProps.foo}baz` };
+    }
+
+    componentDidUpdate(prevProps: Readonly<NewProps>, prevState: Readonly<NewState>, snapshot: { baz: string }) {
+        return;
+    }
+
+    render() {
+        return this.state.bar;
+    }
+}
+<PureComponentWithNewLifecycles foo="bar" />;
 
 class ComponentWithLargeState extends React.Component<{}, Record<'a'|'b'|'c', string>> {
     static getDerivedStateFromProps: React.GetDerivedStateFromProps<{}, Record<'a'|'b'|'c', string>> = () => {

--- a/types/react/test/tsx.tsx
+++ b/types/react/test/tsx.tsx
@@ -150,6 +150,8 @@ class ComponentWithNewLifecycles extends React.Component<NewProps, NewState, { b
         return this.state.bar;
     }
 }
+<ComponentWithNewLifecycles foo="bar" />;
+
 
 class ComponentWithLargeState extends React.Component<{}, Record<'a'|'b'|'c', string>> {
     static getDerivedStateFromProps: React.GetDerivedStateFromProps<{}, Record<'a'|'b'|'c', string>> = () => {

